backend.go

```diff
@@ -255,6 +344,81 @@ func (b *kubeAuthBackend) role(ctx context.Context, s logical.Storage, name stri
 	return role, nil
 }
 
+// getHTTPClient return the backend's HTTP client for connecting to the Kubernetes API.
+func (b *kubeAuthBackend) getHTTPClient(config *kubeConfig) (*http.Client, error) {
+	if b.httpClient == nil {
+		return nil, fmt.Errorf("the backend's http.Client has not been initialized")
+	}
+
+	if b.tlsConfig == nil {
+		// ensure that HTTP client's transport TLS configuration is initialized
+		// this adds some belt-and-suspenders,
+		// since in most cases the TLS configuration would have already been initialized.
+		if err := b.updateTLSConfig(config); err != nil {
+			return nil, err
+		}
+	}
+
+	return b.httpClient, nil
+}
+
+// updateTLSConfig ensures that the httpClient's TLS configuration is consistent
+// with the backend's stored configuration.
+func (b *kubeAuthBackend) updateTLSConfig(config *kubeConfig) error {
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	if b.httpClient == nil {
+		return fmt.Errorf("the backend's http.Client has not been initialized")
+	}
+
+	// attempt to read the CA certificates the config directly or from the filesystem.
+	var caCertBytes []byte
+	if config.CACert != "" {
+		caCertBytes = []byte(config.CACert)
+	} else if !config.DisableLocalCAJwt && b.localCACertReader != nil {
+		// TODO: this may block on I/O, investigate a proper mitigation
+		data, err := b.localCACertReader.ReadFile()
+		if err != nil {
+			return err
+		}
+		caCertBytes = []byte(data)
+	}
+
+	transport, ok := b.httpClient.Transport.(*http.Transport)
+	if !ok {
+		// should never happen
+		return fmt.Errorf("type assertion failed for %T", b.httpClient.Transport)
+	}
+
+	if b.tlsConfig == nil {
+		b.tlsConfig = defaultTLSConfig()
+	}
+
+	certPool := x509.NewCertPool()
+	if len(caCertBytes) > 0 {
+		if ok := certPool.AppendCertsFromPEM(caCertBytes); !ok {
+			b.Logger().Warn("Configured CA PEM data contains no valid certificates, TLS verification will fail")
+		}
+	} else {
+		// provide an empty certPool
+		b.Logger().Warn("No CA certificates configured, TLS verification will fail")
+		// TODO: think about supporting host root CA certificates via a configuration toggle,
+		// in which case RootCAs should be set to nil
+	}
+
+	// only refresh the Root CAs if they have changed since the last full update.
+	if !b.tlsConfig.RootCAs.Equal(certPool) {
+		b.Logger().Trace("Root CA certificate pool has changed, updating the client's transport")
+		b.tlsConfig.RootCAs = certPool
+		transport.TLSClientConfig = b.tlsConfig
```

Comment:
What does setting `transport.TLSClientConfig` do here? It looks like `transport` is local to this function?

Reply:
Oh right, `transport` is a pointer to `b.httpClient.Transport`.